### PR TITLE
Segment compaction: part 1

### PR DESCRIPTION
**Pull request by small-turtle-1**
### What problem does this PR solve?

logic part of segment compaction.
Concurrent control in segment compaction process.
Add unit test for it.

Issue link:
https://github.com/infiniflow/infinity/issues/482
### Code changes

- [x] Has Code change
- [ ] Has CI related scripts change

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (knn performance test)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Note for reviewer

### CODE REVIEW

File: src/common/stl.cppm [hunk 1]
```diff
@@ -264,6 +267,18 @@ export {
         std::transform(str.begin(), str.end(), str.begin(), [](const auto c) { return std::tolower(c); });
     }
 
+    template<class BidirIt>
```
Comment (JinHai-CN) [inline src/common/stl.cppm:270]: BidirIteratorType

File: src/storage/bg_task/compact_segments_task.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+class GreedySegGenerator {
+public:
+    GreedySegGenerator(const Vector<Pair<SegmentID, SizeT>> &segments, SizeT max_segment_size) : max_segment_size_(max_segment_size) {
+        for (const auto &[segment_id, row_count] : segments) {
+            segments_.emplace(row_count, segment_id);
+        }
+    }
+
+    Vector<SegmentID> operator()() {
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:43]: Prefer to use a understandable function name.

File: src/storage/bg_task/compact_segments_task.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+CompactSegmentsTask::CompactSegmentsTask(TableEntry *table_entry, Txn *txn)
+    : BGTask(BGTaskType::kCompactSegments, false), table_entry_(table_entry), txn_(txn) {}
+
+void CompactSegmentsTask::Execute() {
+    Execute1();
+    Execute2();
+}
+
+void CompactSegmentsTask::Execute1() {
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:78]: Use a reasonable name.

File: src/storage/bg_task/compact_segments_task.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        }
+        auto new_segment = CompactSegments(to_compact_segments);
+        segment_data_.emplace_back(new_segment, std::move(to_compact_segments));
+        new_segments_.emplace(new_segment->segment_id(), new_segment);
+    }
+
+    // TODO: Create new index file here
+}
+
+void CompactSegmentsTask::Execute2() {
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:103]: Use a reasonable name.

File: src/storage/bg_task/compact_segments_task.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    auto new_segment = SegmentEntry::NewSegmentEntry(table_entry_, segment_id, txn_);
+
+    TxnTimeStamp begin_ts = txn_->BeginTS();
+    SizeT column_count = table_entry_->ColumnCount();
+    BufferManager *buffer_mgr = txn_->buffer_manager();
+
+    auto &segment_map = table_entry_->segment_map();
+
+    auto *current_block_entry = new_segment->GetLastEntry(); // why segment has at least one block? TODO: remove it
+    UniquePtr<BlockEntry> p = nullptr;
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:163]: single character variable name is hard to understand.

File: src/storage/bg_task/compact_segments_task.cppm [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module;
+
+#include <vector>
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cppm:17]: Avoid to #include headers in cppm file.

File: src/storage/bg_task/compact_segments_task.cppm [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+    const HashMap<BlockID, Vector<BlockOffset>> block_row_hashmap_;
+
+    const TxnTimeStamp commit_ts_;
+};
+
+export class CompactSegmentsTask final : public BGTask {
+public:
+    explicit CompactSegmentsTask(TableEntry *table_entry, Txn *txn);
+
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cppm:85]: virtual destructor?

Comment (small-turtle-1) [reply src/storage/bg_task/compact_segments_task.cppm:85]: If base class destructor is declared as virtual, then the derived class destructor is declared as virtual automatically.

File: src/storage/bg_task/compact_segments_task.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module;
+#include <vector>
+
+import stl;
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:18]: move these import under module declaration.

File: src/storage/bg_task/compact_segments_task.cppm [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    RowID GetNewRowID(RowID old_row_id) const {
+        return GetNewRowID(old_row_id.segment_id_, old_row_id.segment_offset_ / block_capacity_, old_row_id.segment_offset_ % block_capacity_);
+    }
+
+private:
+    const SizeT block_capacity_;
+
+    HashMap<SegmentID, HashMap<BlockID, Vector<Pair<BlockOffset, RowID>>>> row_id_map_;
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cppm:52]: Introduce GlobalBlockID consisted with SegmentID and BlockID

File: src/storage/bg_task/compact_segments_task.cpp [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+class GreedyCompactableSegsGenerator {
+public:
+    GreedyCompactableSegsGenerator(const Vector<Pair<SegmentID, SizeT>> &segments, SizeT max_segment_size) : max_segment_size_(max_segment_size) {
+        for (const auto &[segment_id, row_count] : segments) {
+            segments_.emplace(row_count, segment_id);
+        }
+    }
+
+    Vector<SegmentID> operator()() {
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:62]: Detailed comments

File: src/storage/bg_task/compact_segments_task.cpp [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+CompactSegmentsTask::CompactSegmentsTask(TableEntry *table_entry, Txn *txn)
+    : BGTask(BGTaskType::kCompactSegments, false), table_entry_(table_entry), txn_(txn) {}
+
+void CompactSegmentsTask::Execute() {
+    auto state = CompactSegs();
+    CommitCompacts(std::move(state));
+}
+
+CompactSegmentsTaskState CompactSegmentsTask::CompactSegs() {
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:97]: CompactSegments

File: src/storage/bg_task/compact_segments_task.cpp [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                                                 table_name,
+                                                 *new_segment->segment_dir(),
+                                                 new_segment->segment_id(),
+                                                 block_entries_size,
+                                                 DEFAULT_BLOCK_CAPACITY, // TODO: store block capacity in segment entry
+                                                 last_block_row_count));
+
+        for (SegmentID segment_id : old_segment_ids) {
+            auto *segment_entry = table_entry_->segment_map().at(segment_id).get();
+            segment_entry->SetDeprecated(begin_ts);
```
Comment (JinHai-CN) [inline src/storage/bg_task/compact_segments_task.cpp:225]: These deprecated segment_entry need to be recorded in txn store and WAL. Used in following two cases:
1. Rollbacking the txn will remove the deprecated TS.
2. Replay WAL will deprecated the old segment_entry.

**Review by JinHai-CN — APPROVED** [review]

**Review by small-turtle-1 — COMMENTED** [review]
reviewed.

**Pull request merged**